File: app/orchestrator/limiter.py
import asyncio
import time
from dataclasses import dataclass

from app.storage.sqlite import SQLiteStorage


@dataclass
class LimiterSnapshot:
    """并发限制器状态快照，用于监控排队与占用情况。"""

    max_active: int
    active: int
    waiting: int
    last_wait_ms: float


class RequestLimiter:
    """基于 SQLite 的跨进程并发控制与单用户互斥。"""

    def __init__(
        self,
        storage: SQLiteStorage,
        max_active: int,
        *,
        poll_interval_s: float = 0.2,
        lock_ttl_s: float = 60.0,
    ) -> None:
        self._storage = storage
        self._max_active = max(1, int(max_active))
        self._poll_interval_s = max(0.05, float(poll_interval_s))
        self._lock_ttl_s = max(1.0, float(lock_ttl_s))
        self._lock = asyncio.Lock()
        self._active = 0
        self._waiting = 0
        self._last_wait_ms = 0.0

    @property
    def lock_ttl_s(self) -> float:
        """返回锁的存活时间，供心跳续租使用。"""
        return self._lock_ttl_s

    async def acquire(self, *, session_id: str, user_id: str) -> bool:
        """获取会话执行许可，返回是否成功。"""
        if not user_id or not session_id:
            return False
        waiting_decremented = False
        async with self._lock:
            self._waiting += 1
        start = time.perf_counter()
        try:
            while True:
                result = await asyncio.to_thread(
                    self._storage.try_acquire_session_lock,
                    session_id,
                    user_id,
                    self._max_active,
                    self._lock_ttl_s,
                )
                if result == "acquired":
                    wait_ms = (time.perf_counter() - start) * 1000
                    async with self._lock:
                        self._waiting = max(0, self._waiting - 1)
                        waiting_decremented = True
                        self._active += 1
                        self._last_wait_ms = round(wait_ms, 2)
                    return True
                if result == "user_busy":
                    async with self._lock:
                        self._waiting = max(0, self._waiting - 1)
                        waiting_decremented = True
                    return False
                await asyncio.sleep(self._poll_interval_s)
        finally:
            if not waiting_decremented:
                async with self._lock:
                    self._waiting = max(0, self._waiting - 1)

    async def touch(self, *, session_id: str) -> None:
        """续租会话锁，避免长任务被误判过期。"""
        if not session_id:
            return
        await asyncio.to_thread(
            self._storage.touch_session_lock, session_id, self._lock_ttl_s
        )

    async def release(self, *, session_id: str) -> None:
        """释放会话锁，允许新的请求进入。"""
        if not session_id:
            return
        await asyncio.to_thread(self._storage.release_session_lock, session_id)
        async with self._lock:
            self._active = max(0, self._active - 1)

    def snapshot(self) -> LimiterSnapshot:
        """返回当前状态快照，主要用于本进程内排队观测。"""
        return LimiterSnapshot(
            max_active=self._max_active,
            active=self._active,
            waiting=self._waiting,
            last_wait_ms=self._last_wait_ms,
        )

File: src/orchestrator/limiter.rs
use super::*;
use std::time::{Duration, Instant};

#[derive(Clone)]
pub(super) struct RequestLimiter {
    storage: Arc<dyn StorageBackend>,
    max_active: i64,
    poll_interval_s: f64,
    lock_ttl_s: f64,
}

impl RequestLimiter {
    pub(super) fn new(storage: Arc<dyn StorageBackend>, max_active: usize) -> Self {
        Self {
            storage,
            max_active: max_active.max(1) as i64,
            poll_interval_s: SESSION_LOCK_POLL_INTERVAL_S,
            lock_ttl_s: SESSION_LOCK_TTL_S,
        }
    }

    pub(super) async fn acquire(
        &self,
        session_id: &str,
        user_id: &str,
        agent_id: &str,
        allow_queue: bool,
    ) -> Result<bool> {
        let cleaned_session = session_id.trim();
        let cleaned_user = user_id.trim();
        let cleaned_agent = agent_id.trim();
        if cleaned_session.is_empty() || cleaned_user.is_empty() {
            return Ok(false);
        }
        if !allow_queue {
            let storage = self.storage.clone();
            let session_id = cleaned_session.to_string();
            let user_id = cleaned_user.to_string();
            let agent_id = cleaned_agent.to_string();
            let ttl = self.lock_ttl_s;
            let max_active = self.max_active;
            let status = tokio::task::spawn_blocking(move || {
                storage.try_acquire_session_lock(&session_id, &user_id, &agent_id, ttl, max_active)
            })
            .await
            .map_err(|err| anyhow!("session lock join error: {err}"))??;
            return Ok(matches!(status, SessionLockStatus::Acquired));
        }
        let retry_window = SESSION_LOCK_BUSY_RETRY_S.max(self.poll_interval_s);
        let retry_deadline = Instant::now() + Duration::from_secs_f64(retry_window);
        loop {
            let storage = self.storage.clone();
            let session_id = cleaned_session.to_string();
            let user_id = cleaned_user.to_string();
            let agent_id = cleaned_agent.to_string();
            let ttl = self.lock_ttl_s;
            let max_active = self.max_active;
            let status = tokio::task::spawn_blocking(move || {
                storage.try_acquire_session_lock(&session_id, &user_id, &agent_id, ttl, max_active)
            })
            .await
            .map_err(|err| anyhow!("session lock join error: {err}"))??;
            match status {
                SessionLockStatus::Acquired => return Ok(true),
                SessionLockStatus::UserBusy => {
                    if Instant::now() >= retry_deadline {
                        return Ok(false);
                    }
                    tokio::time::sleep(Duration::from_secs_f64(self.poll_interval_s)).await;
                }
                SessionLockStatus::SystemBusy => {
                    tokio::time::sleep(Duration::from_secs_f64(self.poll_interval_s)).await;
                }
            }
        }
    }

    pub(super) async fn touch(&self, session_id: &str) {
        let cleaned_session = session_id.trim();
        if cleaned_session.is_empty() {
            return;
        }
        let storage = self.storage.clone();
        let session_id = cleaned_session.to_string();
        let ttl = self.lock_ttl_s;
        match tokio::task::spawn_blocking(move || storage.touch_session_lock(&session_id, ttl))
            .await
        {
            Ok(Ok(())) => {}
            Ok(Err(err)) => {
                warn!("failed to touch session lock for {cleaned_session}: {err}");
            }
            Err(err) => {
                warn!("failed to touch session lock for {cleaned_session}: {err}");
            }
        }
    }

    pub(super) async fn release(&self, session_id: &str) {
        let cleaned_session = session_id.trim();
        if cleaned_session.is_empty() {
            return;
        }
        let storage = self.storage.clone();
        let session_id = cleaned_session.to_string();
        match tokio::task::spawn_blocking(move || storage.release_session_lock(&session_id)).await {
            Ok(Ok(())) => {}
            Ok(Err(err)) => {
                warn!("failed to release session lock for {cleaned_session}: {err}");
            }
            Err(err) => {
                warn!("failed to release session lock for {cleaned_session}: {err}");
            }
        }
    }
}
